src/app/compare/Compare.tsx

```diff
@@ -185,6 +191,121 @@ class Compare extends React.Component<Props, State> {
 
     const stats = this.getAllStatsSelector(this.state, this.props);
 
+    // TODO: Should this go into its own module?
+    const updateSocketComparePlug = ({
+      item,
+      categoryHash,
+      socket,
+      plug: clickedPlug,
+    }: {
+      item: DimItem;
+      categoryHash: number;
+      socket: DimSocket;
+      plug: DimPlug;
+    }) => {
+      // Exit early if this is a D1 item or an item without sockets
+      if (!item.isDestiny2() || !item.sockets || item.sockets === undefined) {
+        return null;
+      }
+
+      let { socketIndex } = socket;
+      const categoryIndex = item.sockets.categories.findIndex(
+        (category) => category.category.hash === categoryHash
+      );
+      // Create deep copies of arrays to prevent directly mutating state
+      // (cloneDeep is resulting in object mutation)
+      const itemStatsClone: DimStat[] = JSON.parse(JSON.stringify(item.stats));
+
+      // socketIndex is correct for allSockets, but not categories[].sockets
+      // TODO: Confirm socket is always off by 1, otherwise this should use hashes to find plugs
+      socketIndex -= 1;
+
+      // Exit early if the socket index doesn't exist (kill tracker sockets)
+      if (!item.sockets.categories[categoryIndex].sockets[socketIndex]) {
+        return null;
+      }
+
+      if (
+        (adjustedPlugs?.[item.id]?.[socketIndex] === undefined &&
+          clickedPlug.plugDef.hash !==
+            item.sockets.categories[categoryIndex].sockets[socketIndex].plugged?.plugDef.hash) ||
+        clickedPlug.plugDef.hash !== adjustedPlugs?.[item.id]?.[socketIndex]?.plugDef.hash
+      ) {
+        // Clone stats so state isn't directly mutated by math
+        const updateAdjustedPlugs =
+          isEmpty(adjustedPlugs) || adjustedPlugs === undefined
+            ? { [item.id]: {} }
+            : adjustedPlugs[item.id]
+            ? cloneDeep(adjustedPlugs)
```

Comment:
Can you have a look at using immer's produce over cloneDeep? It will avoid have to to assign things and has been our pattern for updating state without mutations.

Reply:
Thanks for the suggestion - I had a bit of a fight with `produce` handling undefined objects, so it would be handy if we could go over that at some point.
`produce(adjustedPlugs ?? {}, (draft) => { draft[item.id] = { [socketIndex]: nextPlug }; })`